### ISSUE TITLE
Traversal: Improve traversal performance

### ISSUE THREAD

**Issue by gkjohnson** [NASA-AMMOS/3DTilesRendererJS#678]
Related to #669

- Merge some of the functions. Like at least `markVisibleTiles` and `markUsedSetLeaves`.
- Accumulate toggled tiles into a set or array so it can be iterated once rather than unnecessarily traversing through all parents.
- Lazily generate the error value & frustum state as-needed since it isn't always used.

Related goal: improve performance of loading / traversing tile set from https://github.com/iTowns/itowns/issues/2335

**Comment by pmi-gbs** [comment]
We're still (related #1006) suffering from this a lot, basically half work or more for each frame is spent in this:

<img width="2071" height="1025" alt="Image" src="https://github.com/user-attachments/assets/7483498e-a7e9-402a-bb3a-cdc94a864705" />

I'd like to tackle this...

Apart of the suggestions in the first comment, I wonder whether it's worth adding a separate web-worker, transferring minimal updates (deltas).
This is likely a bit of extra book-keeping I think, but it would be great to free the main thread from all work that can be offloaded somehow...

Main issue I currently see for this, is that there's currently a hook (`calculateTileViewErrorWithPlugin`) for a plugin in `resetFrameState` (one of the main blockers)

**Comment by gkjohnson** [comment]
Some help making improvements would be great. I expect there are some perf improvements to be made in the traversal (like not checking error / frustum for children where the parent did not pass). But probably not anything that will get the timing down as low as you want for this kind of use case - I'd think the biggest limitation is the depth the tree is being traversed to.

Offloading traversal to a web worker sounds like it would be a benefit but they bring a lot of complexity (for both implementation and users) and will inherently have a delay so I don't think I'm ready to say we should refactor the project to be web-worker first. I'm open to any outlines around how this could work. But here are a few ideas that come to mind from least to most invasive:

### Generator Traversal w/ Deadline

The traversal functions could be converted to a series of generators that yield based on a provided deadline. The traversal would be completed in subsequent frames and tile visibility toggles would not fire until the traversal finished completely. This would mean you could set the per-frame budget to something like 2ms per frame so the traversal would finish after 7 or 8 frames in your example.

This is something that could be added fairly simply and bring a noticeable benefit.

### Move the TilesRenderer to Worker

Instead of instantiating the TilesRenderer instance on the main thread, it would be created in a Worker thread with a thin wrapper in the main thread to proxy necessary data and events through. This would mean the camera and resolution values would need to be passed to the worker and likewise any "load tile" or "tile visible" events would need to be passed to and dispatched through the thin main thread wrapper to toggle visibility, etc.

This would make some plugins and anything else that requires hierarchy inspection difficult or impossible to use but this could probably already be implemented with a wrapper or plugin and would be the most simple way to get a version of this working.

### Traversal-only Worker

Ideally we'd just be able to run the traversal on a web worker while the main TilesRenderer instance remains accessible on the main thread and processes any of the traversal deltas (toggled tiles, required downloads). But this would mean we need a way of synchronizing the hierarchy in both places (and that's double the memory for the hierarchy) if there are any plugins that require access - applying "fixups" from plugins like ImplicitTilingPlugin or QuanitizedMeshPlugin bring added complexity, for example.

**Comment by pmi-gbs** [comment]
Thanks very much for the detailed suggestions, I'll take a look at it in some time (need to take a break currently because of tendinitis).

The deadline approach sounds good as a first approach, to see whether that maybe is already enough (including maybe general optimizations as well, though I've seen it's already fairly optimized).

What may also be interesting to look at is something like probabilistic traversal heuristics, i.e. instead of traversing the whole tree, try to prioritize traversal in the most likely candidates that need refinement (or unloading) within a deadline.

Though just some thoughts.
I need to dive deeper in the code-base to see whether something like that is even viable (and how complex a worker thread really is, I'm mostly concerned about external API-surface currently, when it's mostly limited to internal plugins then I think it's not the biggest issue, as I'm sure they can be adjusted).

**Comment by pmi-gbs** [comment]
> and will inherently have a delay

btw. I did recently implement a SVO based streamed SDF-Renderer (Deviation Analysis), that uses a worker for generating a histogram of hit subtrees (encoded subtree-paths rasterized by GPU) by counting pixels (in Rust because it was 10 times faster than in JS, also because the implementation could be much more optimized, using an optimized hash function on encoded octree-path-indices). I originally wanted to use the GPU, but well... WebGL2 is unfortunately extremely limited, so I had to do that on the CPU (I can't await finally switching to WebGPU with proper compute shaders, when browser support is *really* stable)

Long story short, I think it works pretty well, the delay is barely if at all noticeable (well roughtly 1-4 frames usually, depending on the amount of time the worker takes). I did only transfer limited update requests back to the main thread (and from the main thread the rasterized encoded texture to the worker)
